### ISSUE TITLE
group support

### ISSUE THREAD

**Issue by dweinstein** [tomas-abrahamsson/gpb#106]
@tomas-abrahamsson,

many (albeit internal) google APIs use the deprecated groups feature in protobuf definitions. is there any feeling against supporting this deprecated feature of protobuf? Would it be difficult to introduce support in gpb, or would it introduce other issues that would impact the current functionality? Thanks...

**Comment by tomas-abrahamsson** [comment]
I'll look into it

**Comment by tomas-abrahamsson** [comment]
FYI: I think I'm going to do it like this: If we have a message like this
```proto
message m1 {
  required group g = 2 {
    required uint32 gf = 3;
  }
}
```
...then the name of group will be `'m1.m1_g'`, and there will be a message "invented" for it. The name will not be `m1_g` because that could possibly collide with some other message definition. Naming it `'m1.m1_g'` is safer, I think.

On technical details, I think I will make the internal structure be `{{group,Name},Fields}` and the field type for such a group (field type for the field `g` in the example) will be `{group,Name}`. (The internal structure is exposed via [the `to_proto_defs` compilation option](https://github.com/tomas-abrahamsson/gpb/blob/3.26.8/src/gpb_compile.erl#L293-L296).)

It is not carved in stone yet, I might change my mind later, but for now, I think this is how it will look.

**Comment by dweinstein** [comment]
sounds a little like https://github.com/dcodeIO/protobuf.js/issues/568#issuecomment-267795981

**Comment by tomas-abrahamsson** [comment]
Yes, similar

**Comment by tomas-abrahamsson** [comment]
Hi, I've just pushed a branch, `groups` if you want to try it out. Let me know how it works for you, and I can merge it to master. Still missing on this branch is [nif support](https://github.com/tomas-abrahamsson/gpb/blob/3.26.8/README.nif-cc), but I can add that later.

I changed my mind about naming, so the group in the previous example would be just `'m1.g'` not `'m1.m1_g'`. As I learned more while implementing groups, I no longer think there's any risk of naming collision.
